### PR TITLE
[FLINK-24385][table] Introduce TRY_CAST

### PR DESCRIPTION
**Pull request by slinkydeveloper**
## What is the purpose of the change

The goal of this PR is to

* Let CAST fail
* Introduce TRY_CAST to handle cast failures
* When legacy cast behavior is enabled, restore CAST behavior to not fail

## Brief change log

* Improved the `canCast` method in the `CastRule` stack to do proper nested checking of `canFail`.
* `CAST` now fails if legacy cast is disabled, and `canFail` returns true. When legacy cast is enabled, the cast logic is wrapped in the codegen with a try-catch block
* Implemented `TRY_CAST` in SQL. This includes: 
  * changes to the parser (no new custom sql node, as it's not needed)
  * new `SqlTryCastFunction` mimicking the behavior of `SqlCastFunction` from calcite
  * add a custom `SqlConvertletTable` to plug in the conversion of the `TRY_CAST` operator (this is necessary as in the parser we cannot access `FlinkSqlOperatorTable.TRY_CAST`)
* Implemented `TRY_CAST` in Table API. This includes:
  * Add `TRY_CAST` to `BuiltInFunctionDefinition`
  * Converter back and forth the `Expression` stack
* Use the legacy cast behavior flag in to enable/disable

## Verifying this change

This PR adds test cases in `CastFunctionMiscITCase` for testing the new `TRY_CAST`. Relies on previous existing tests for checking the behavior of `CAST` in legacy mode.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes, **no breaking change when legacy mode is enabled**

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
## Some implementation notes

Below some implementation details worth to be mentioned

### Why `SqlTryCastFunction` is `SqlKind.OTHER_FUNCTION`

Initially I tried to use `SqlKind.CAST` for `SqlTryCastFunction`, as it effectively unlocks a couple of nice features from calcite, notably several optimizations and cast removals. This unfortunately also comes with some implicit assumptions done by calcite, after checking that the call is of `SqlKind.CAST`.

In particular, I hit a wall with `RexSimplify`, which was trying to optimize `TRY_CAST('non-numeric' AS BIGINT)` in a test. Because the value to cast is a literal, calcite tried to cast it at planning time, failing rather than converting the whole expression to `NULL`. Fixing this issue requires me to hardcode some custom logic in `RexSimplify`, which is not extensible.

### Implicit casting doesn't fail

IMHO this is a bigger issue than this PR, as the planner shouldn't try at all to perform implicit casting among types which casting can fail. To avoid substantial differences, I kept the situation as it is, but I suggest we go back on this issue later in future as we should disallow such implicit castings. 

### How the old legacy behavior is still supported

Initially my idea to support the legacy behavior was to implement a planner rule converting every CAST to TRY_CAST, in case legacy behavior is enabled.

The problem with this approach is that converting to `TRY_CAST` during planning entails modifying the result type of the expression, eventually generating a breaking change anyway. For example, given a column defined as `f0 STRING NOT NULL` and the query `SELECT CAST(f0 AS TIME) FROM T`, the plan generated from master would be:

```
LogicalProject(EXPR$0=[CAST($0):TIME(0) NOT NULL])
``` 

But now, because `STRING` to `TIME` can fail, the plan should rather be:

```
LogicalProject(EXPR$0=[TRY_CAST($0):TIME(0)])
``` 

Such change effectively break the plans, as now a different nullability constraints can change the outcome of the query planning. Hence the user should handle this null as it would be with `TRY_CAST`. e.g. think about a function that accepts NOT NULL parameters and its argument is the above cast, that would not compile anymore with this change without user intervention to check the nullability, for example wrapping the cast in a `COALESCE`.

On top of that, Calcite doesn't allow to modify the output type of a `RelNode` inside rules (check `HepRuleCall#transformTo`), so modifying the type requires to perform this call rewrite at the SqlNode/Expression -> RexNode level, rather than in a planner rule.

Then there's also the issue with implicit castings, placed here and there not just by our rules, but also by calcite various components'.

Because of the aforementioned problems, It's quite complicated to write a clean rule that does the job, which made me fallback to the _quick and dirty_ fix directly in the codegen.

It must be noted that, while on one hand having a rule nicely converting `CAST` to `TRY_CAST` changes the plan and constitutes a breaking change, it effectively makes the upgrade path for future Flink versions easier, as such rule effectively adapt the `CAST` semantics as if the user just used `TRY_CAST`.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionMiscITCase.java [hunk 1]
```diff
@@ -206,7 +209,48 @@ protected Configuration configuration() {
...
+                TestSpec.forFunction(
+                                BuiltInFunctionDefinitions.TRY_CAST,
+                                "try cast from TIME NOT NULL to STRING NOT NULL")
+                        .onFieldsWithData(LocalTime.parse("12:34:56"))
+                        .andDataTypes(TIME().notNull())
+                        .testResult(
+                                $("f0").tryCast(STRING()),
+                                "TRY_CAST(f0 AS STRING)",
+                                "12:34:56",
+                                // Because TIME to STRING cannot fail, the type inference should
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionMiscITCase.java:236]:
```suggestion
                                // Because TIME to STRING cannot fail, the type inference should be
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java [hunk 2]
```diff
@@ -144,58 +143,69 @@ private static void testResult(
...
-        } catch (Throwable t) {
-            assertTrue(t instanceof ValidationException);
-            assertThat(t.getMessage(), containsString(testItem.errorMessage));
+        AtomicReference<TableResult> tableResult = new AtomicReference<>();
+
+        Throwable t =
+                catchThrowable(
+                        () -> {
+                            if (testItem instanceof TableApiErrorTestItem) {
+                                tableResult.set(
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:179]: This and the else block, maybe extract them to methods to improve readability.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java [hunk 3]
```diff
@@ -144,58 +143,69 @@ private static void testResult(
...
-                    result.getResolvedSchema().getColumnDataTypes().get(i).getLogicalType());
-
-            assertEquals(
-                    "Result for spec [" + i + "] of test [" + testItem + "] doesn't match.",
-                    // Use Row.equals() to enable equality for complex structure, i.e. byte[]
-                    Row.of(testItem.results.get(i)),
-                    Row.of(row.getField(i)));
+            assertThat(result.getResolvedSchema().getColumnDataTypes().get(i).getLogicalType())
+                    .as(
+                            "Logical type for spec ["
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:155]: nit: I would try to compact this to less line breaks.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java [hunk 4]
```diff
@@ -33,7 +39,11 @@ public CastRulePredicate getPredicateDefinition() {
     }
 
     @Override
-    public boolean canFail() {
-        return false;
+    public boolean canFail(LogicalType inputLogicalType, LogicalType targetLogicalType) {
+        final List<LogicalType> inputFields = LogicalTypeChecks.getFieldTypes(inputLogicalType);
+        final List<LogicalType> targetFields = LogicalTypeChecks.getFieldTypes(targetLogicalType);
+
+        return IntStream.range(0, Math.min(inputFields.size(), targetFields.size()))
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java:46]: To my understanding this is not correct, don't we need to check all the combinations of each of the inputFields with all the targetFields (nested loop)?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkConvertletTable.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        } else {
+            throw new IllegalStateException(
+                    "Invalid right argument type for TRY_CAST: " + rightNode);
+        }
+
+        final LogicalType fromLogicalType = FlinkTypeFactory.toLogicalType(valueRex.getType());
+        final LogicalType toLogicalType = FlinkTypeFactory.toLogicalType(type);
+
+        // This is nullable only and only if the cast rule can fail
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkConvertletTable.java:86]: just `only` would be enough I think.

File: docs/data/sql_functions.yml [hunk 6]
```diff
@@ -561,7 +561,10 @@ conditional:
 conversion:
   - sql: CAST(value AS type)
     table: ANY.cast(TYPE)
-    description: Returns a new value being cast to type type. E.g., CAST('42' AS INT) returns 42; CAST(NULL AS VARCHAR) returns NULL of type VARCHAR.
+    description: Returns a new value being cast to type type. A CAST error throws an exception and fails the job. If you're performing a cast operation that may fail, like INT to STRING, you should rather use TRY_CAST, in order to handle errors. E.g., CAST('42' AS INT) returns 42; CAST(NULL AS VARCHAR) returns NULL of type VARCHAR; TRY_CAST('non-number' AS INT) throws an exception and fails the job.
```
Comment (matriv) [inline docs/data/sql_functions.yml:564]: I think we need to refer to the LEGACY BEHAVIOUR config param in `execution_config_configuration.html` to be 100% clear to the user.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java [hunk 7]
```diff
@@ -1245,13 +1253,23 @@ private CastTestSpecBuilder fail(DataType dataType, Object src) {
...
+        private CastTestSpecBuilder failRuntime(
+                DataType dataType, Object src, Class<? extends Throwable> failureClass) {
+            this.testTypes.add(TestType.ERROR_RUNTIME);
+            this.columnTypes.add(dataType);
+            this.columnData.add(src);
+            this.expectedValues.add(failureClass);
+            return this;
+        }
+
         private CastTestSpecBuilder fail(TestType type, DataType dataType, Object src) {
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:1265]: Maybe rename this to `failValidation`?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java [hunk 8]
```diff
@@ -1245,13 +1253,23 @@ private CastTestSpecBuilder fail(DataType dataType, Object src) {
...
             this.testTypes.add(type);
             this.columnTypes.add(dataType);
             this.columnData.add(src);
             return this;
         }
 
+        @SuppressWarnings("unchecked")
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java:1272]: nit: maybe move it on top of the statements with the cast to `Class<? extends Throwable>`

**Review by matriv — COMMENTED** [review]
Thank you very much @slinkydeveloper!
Great work and great description of the findings and thought process!

Left some comments.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java [hunk 9]
```diff
@@ -117,7 +117,7 @@ private FlinkAssertions() {}
      *                  .hasMessageContaining(containsMessage));
      * }</pre>
      */
-    public static ThrowingConsumer<? extends Throwable> anyCauseMatches(String containsMessage) {
+    public static ThrowingConsumer<? super Throwable> anyCauseMatches(String containsMessage) {
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java:120]: that's the signature accepted by `satisfies` in assertj. Check the above method

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRuleProviderTest.java [hunk 10]
```diff
@@ -77,4 +82,18 @@ void testResolveConstructedToString() {
         assertThat(CastRuleProvider.resolve(new ArrayType(INT), new VarCharType(10)))
                 .isSameAs(ArrayToStringCastRule.INSTANCE);
     }
+
+    @Test
+    void testCanFail() {
+        LogicalType inputType = ROW(INT(), STRING()).getLogicalType();
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRuleProviderTest.java:88]: but `INT` and `TINYINT` variables are `LogicalType`s, not `DataType`s

[the remark this answers is not in the document]
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/casting/CastRuleProviderTest.java:88]: yep yep, sorry, my bad.

File: docs/data/sql_functions.yml (hunk 6 above)
Comment (slinkydeveloper) [reply docs/data/sql_functions.yml:564]: TBH i don't know how to add a link here, but in any case, I've added more text.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java (hunk 4 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java:46]: So this is a generic logic that is supposed to work with every coimbination of types that has children, that is row to row, map to map, array to array and multiset to multiset. For types without children, it always returns false.

So more than correct i would say this works for these cases, for the others, you need to override this method yourself. Does it makes sense?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 11]
```diff
@@ -476,6 +477,15 @@ public OutType cast(DataType toType) {
         return toApiSpecificExpression(unresolvedCall(CAST, toExpr(), typeLiteral(toType)));
     }
 
+    /**
+     * Converts a value to a given data type.
+     *
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:482]: I've synced this doc with the one from sql_functions.yml

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java (hunk 4 above)
Comment (matriv) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java:46]: yep yep, again, I lost the context. all good.
Should there be a check though that both lists have the same size? is there a case that we actually need the `Math.min()`?

File: docs/data/sql_functions.yml [hunk 12]
```diff
@@ -561,7 +561,7 @@ conditional:
 conversion:
   - sql: CAST(value AS type)
     table: ANY.cast(TYPE)
-    description: Returns a new value being cast to type type. A CAST error throws an exception and fails the job. If you're performing a cast operation that may fail, like INT to STRING, you should rather use TRY_CAST, in order to handle errors. E.g., CAST('42' AS INT) returns 42; CAST(NULL AS VARCHAR) returns NULL of type VARCHAR; TRY_CAST('non-number' AS INT) throws an exception and fails the job.
+    description: Returns a new value being cast to type type. A CAST error throws an exception and fails the job. If you're performing a cast operation that may fail, like INT to STRING, you should rather use TRY_CAST, in order to handle errors. If "table.exec.sink.legacy-cast-behaviour" is enabled, CAST behaves like TRY_CAST. E.g., CAST('42' AS INT) returns 42; CAST(NULL AS VARCHAR) returns NULL of type VARCHAR; CAST('non-number' AS INT) throws an exception and fails the job.
```
Comment (matriv) [inline docs/data/sql_functions.yml:564]: Just a reminder to remove `sink` when the option is renamed.

**Review by matriv — APPROVED** [review]
LGTM. Thank you @slinkydeveloper !

File: docs/data/sql_functions.yml (hunk 12 above)
Comment (slinkydeveloper) [reply docs/data/sql_functions.yml:564]: Let me do it in this PR then

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java (hunk 4 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java:46]: The thing is that you don't reach this code if the rule is not matched first, hence i don't think we need to be strict here with checks.

Comment (matriv) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java:46]: Just saying, that for someone seeing this code in the future, might be confusing.
If we check for same size, we don't need to do the `Math.min()` as well, and we can recognise some change that breaks this size equality.

Comment (matriv) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java:46]: throw an exception, again, I'm not insisting at all, just as a "safety net".

Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java]: Ok, I modified this code again, now this logic is in `CollectionToCollectionCastRule` and then `RowToRowCastRule` has its custom one. Check it again

Comment (matriv) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AbstractCastRule.java]: nice! I like it this way, thank you!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/RowToRowCastRule.java [hunk 13]
```diff
@@ -224,4 +224,13 @@ protected String generateCodeBlockInternal(
         writer.stmt(methodCall(writerTerm, "complete")).assignStmt(returnVariable, rowTerm);
         return writer.toString();
     }
+
+    @Override
+    public boolean canFail(LogicalType inputLogicalType, LogicalType targetLogicalType) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/RowToRowCastRule.java:229]: implement `CollectionToCollectionCastRule` here as well?

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java (hunk 9 above)
Comment (twalthr) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java:120]: put this into a separate commit, because it is a core change

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 14]
```diff
@@ -1529,6 +1529,20 @@ public void testSetReset() {
...
+    public void testTryCast() {
+        // Note that is expected that the unparsed value has the comma rather than AS, because we
+        // don't use a custom SqlNode for TryCast, but we rely on SqlBasicCall
+
+        // Simple types
+        expr("try_cast(a as timestamp)").ok("TRY_CAST(`A` AS TIMESTAMP)");
+        expr("try_cast('abc' as timestamp)").ok("TRY_CAST('abc' AS TIMESTAMP)");
+
+        // Complex types
+        expr("try_cast(a as row(f0 int, f1 varchar))")
```
Comment (twalthr) [inline flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:1542]: this is a type from Calcite, can we also try Flink complex types e.g.
`TRY_CAST(f0 AS ROW<a ARRAY<INT>, b MAP<STRING, DECIMAL(10, 2)>>)`

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 15]
```diff
@@ -1529,6 +1529,20 @@ public void testSetReset() {
         sql("RESET 'test-key'").ok("RESET 'test-key'");
     }
 
+    @Test
+    public void testTryCast() {
+        // Note that is expected that the unparsed value has the comma rather than AS, because we
```
Comment (twalthr) [inline flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:1534]: is this comment still valid? I don't see a comma in this test?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkConvertletTable.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+            type = dataType.deriveType(cx.getValidator());
+            if (type == null) {
+                type = cx.getValidator().getValidatedNodeType(dataType.getTypeName());
+            }
+        } else {
+            throw new IllegalStateException(
+                    "Invalid right argument type for TRY_CAST: " + rightNode);
+        }
+
+        final LogicalType fromLogicalType = FlinkTypeFactory.toLogicalType(valueRex.getType());
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/calcite/FlinkConvertletTable.java:83]: Let's simplify the logic in the stack. It is ok if `TRY_CAST` always returns NULLABLE. Users can simply use `CAST` if they don't expect an error. It also makes sure that their is no mismatch between Table API type inference and SQL.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlTryCastFunction.java [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+     */
+    SqlTryCastFunction() {
+        super(
+                "TRY_CAST",
+                DEFAULT_VERSION,
+                SqlKind.OTHER_FUNCTION,
+                null,
+                SqlStdOperatorTable.CAST
+                        .getOperandTypeInference(), // From Calcite's SqlCastFunction
+                null,
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlTryCastFunction.java:60]: We should try to have a complete copy of `CAST`. E.g. `CAST` defines `InferTypes.FIRST_KNOWN` also the monotonicity property can be adopted.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlTryCastFunction.java [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        }
+        call.operand(1).unparse(writer, 0, 0);
+        writer.endFunCall(frame);
+    }
+
+    @Override
+    public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
+        RelDataType returnType = SqlStdOperatorTable.CAST.inferReturnType(opBinding);
+
+        final LogicalType fromLogicalType =
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlTryCastFunction.java:105]: as mentioned before, let's simplify the logic and don't switch to Flink's type system.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/converters/TryCastConverter.java [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        final TypeLiteralExpression targetType = (TypeLiteralExpression) call.getChildren().get(1);
+
+        final LogicalType fromType = FlinkTypeFactory.toLogicalType(child.getType());
+        final LogicalType toType = targetType.getOutputDataType().getLogicalType();
+
+        // We need to adjust the type nullability here, as in table-common we cannot implement it
+        // correctly because we cannot access CastRuleProvider#canFail
+        RelDataType targetRelDataType =
+                context.getTypeFactory().createFieldTypeFromLogicalType(toType);
+        if (CastRuleProvider.canFail(fromType, toType)) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/converters/TryCastConverter.java:60]: both casts can share the same parameterized `CastConverter`

File: flink-python/pyflink/table/expression.py [hunk 20]
```diff
@@ -837,12 +837,28 @@ def alias(self, name: str, *extra_names: str) -> 'Expression[T]':
 
     def cast(self, data_type: DataType) -> 'Expression':
         """
-        Converts a value to a given data type.
+        Returns a new value being cast to type type.
+        A cast error throws an exception and fails the job.
+        If you're performing a cast operation that may fail, like INT to STRING,
```
Comment (twalthr) [inline flink-python/pyflink/table/expression.py:842]: for the future: use `one` instead of `you`, or no person like: `When performing a cast operation that may fail, use...`

File: docs/data/sql_functions.yml (hunk 6 above)
Comment (twalthr) [inline docs/data/sql_functions.yml]: remove `TRY_`

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 21]
```diff
@@ -469,18 +469,28 @@ public OutType collect() {
     }
 
     /**
-     * Converts a value to a given data type.
+     * Returns a new value being cast to {@code toType}. A cast error throws an exception and fails
+     * the job. If you're performing a cast operation that may fail, like {@link DataTypes#INT()} to
+     * {@link DataTypes#STRING()}, you should rather use {@link #tryCast(DataType)}, in order to
+     * handle errors. If {@code table.exec.sink.legacy-cast-behaviour} is enabled, this function
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java]: link to config option directly

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for this PR @slinkydeveloper. I just found one issue that we should address before merging. Otherwise the comments are rather minor.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/RowToRowCastRule.java (hunk 13 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/RowToRowCastRule.java:229]: No, because the logic is not the same: the row to row cast rule allows for target type to have less fields than the source type, so we cannot assert that the children type list has the same size across the two types. Look a couple of lines below when i use `Math.min` and also look at the `matches` method implementation.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java (hunk 9 above)
Comment (slinkydeveloper) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/FlinkAssertions.java:120]: +1 just as FYI no one is using this method (also because without this change is unusable :smile: )

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java (hunk 14 above)
Comment (slinkydeveloper) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:1542]: Added this specific case

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlTryCastFunction.java (hunk 17 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlTryCastFunction.java:60]: And in fact we're adopting it, without copying the actual implementation, but rather simply using the getters to get monotonicity and operand type inference. it's essentially no different than directly using `InferTypes.FIRST_KNOWN` here, with the difference that this potentially is one less pain to fix when upgrading to the next calcite version

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/converters/TryCastConverter.java (hunk 19 above)
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/expressions/converter/converters/TryCastConverter.java:60]: After removing this code that was simplifying the not fallible case, not really, as this code now only differs for very specific sql or table api calls for inferring the return type

File: docs/data/sql_functions.yml (hunk 6 above)
Comment (slinkydeveloper) [reply docs/data/sql_functions.yml]: This has been fixed already in one of the last commits of this PR

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java [hunk 22]
```diff
@@ -107,6 +108,16 @@ public static boolean exists(LogicalType inputType, LogicalType targetType) {
...
     }
 
+    /**
+     * Resolves the rule and returns the result of {@link CastRule#canFail(LogicalType,
+     * LogicalType)}. Fails with {@link NullPointerException} if the rule cannot be resolved.
+     */
+    public static boolean canFail(LogicalType inputType, LogicalType targetType) {
+        return Preconditions.checkNotNull(
+                        resolve(inputType, targetType), "Cast rule cannot be resolved")
+                .canFail(inputType, targetType);
```
[the remark this answers is not in the document]
Comment (slinkydeveloper) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleProvider.java]: I pushed in the last commit a consistent change to fix this, in order to merge the "can fail" algorithm with the rule matching algorithm.

**Review by twalthr — APPROVED** [review]
I found one minor thing. We should check once more if all doc pages are synchronized with the latest version. For the last commit, I would ask @matriv who might have more insights in the cast rule stack.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 23]
```diff
@@ -468,14 +470,33 @@ public OutType collect() {
...
     }
 
+    /**
+     * Like {@link #cast(DataType)}, but in case of error, returns {@code null} rather than failing
+     * the job.
+     *
+     * <p>E.g. {@code "42".tryCast(DataTypes.INT())} returns {@code 42}; {@code
+     * null.tryCast(DataTypes.STRING())} returns {@code null} of type {@link DataTypes#STRING()};
+     * {@code "non-number".tryCast(DataTypes.INT())} returns {@code null} of type {@link
+     * DataTypes#INT()}.
```
Comment (matriv) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java:494]: How about suggesting the use of `coalesce` to wrap the `tryCast` and  provide a different default value than `null`.

File: flink-python/pyflink/table/expression.py [hunk 24]
```diff
@@ -837,12 +837,28 @@ def alias(self, name: str, *extra_names: str) -> 'Expression[T]':
...
         """
         return _binary_op("cast")(self, _to_java_data_type(data_type))
 
+    def try_cast(self, data_type: DataType) -> 'Expression':
+        """
+        Like cast, but in case of error, returns NULL rather than failing the job.
+
+        E.g. lit("42").try_cast(DataTypes.INT()) returns 42;
+        lit(null).try_cast(DataTypes.STRING()) returns NULL of type STRING;
+        list("non-number").cast(DataTypes.INT()) returns NULL of type INT.
```
Comment (matriv) [inline flink-python/pyflink/table/expression.py:858]: same here, mentioning `coalesce`.

File: docs/data/sql_functions.yml [hunk 25]
```diff
@@ -561,7 +561,10 @@ conditional:
 conversion:
   - sql: CAST(value AS type)
     table: ANY.cast(TYPE)
-    description: Returns a new value being cast to type type. E.g., CAST('42' AS INT) returns 42; CAST(NULL AS VARCHAR) returns NULL of type VARCHAR.
+    description: Returns a new value being cast to type type. A CAST error throws an exception and fails the job. When performing a cast operation that may fail, like STRING to INT, one should rather use TRY_CAST, in order to handle errors. If "table.exec.legacy-cast-behaviour" is enabled, CAST behaves like TRY_CAST. E.g., CAST('42' AS INT) returns 42; CAST(NULL AS STRING) returns NULL of type STRING; CAST('non-number' AS INT) throws an exception and fails the job.
+  - sql: TRY_CAST(value AS type)
+    table: ANY.tryCast(TYPE)
+    description: Like CAST, but in case of error, returns NULL rather than failing the job. E.g., TRY_CAST('42' AS INT) returns 42; TRY_CAST(NULL AS STRING) returns NULL of type STRING; TRY_CAST('non-number' AS INT) returns NULL of type INT.
```
Comment (matriv) [inline docs/data/sql_functions.yml:567]: Maybe also here, mention wrapping with `coalesce` for different default value.

**Review by matriv — APPROVED** [review]
Thank you @slinkydeveloper 

 - I added some suggestions to mention coalesce on top of try cast.
 - Honestly, I don't like the last commit with the refactoring. I think it complicates things too much compared to the problem it solves. On top, I don't like mixing `boolean` with `CastRuleMatch`, and that we don't use `.is(LogigalTypeRoot(xxx)` but we use the concrete class. Imho, I'd remove this change for this PR, and try to rethink the approach in a followup.

**Comment by slinkydeveloper** [comment]
@matriv removed the last commit and added the coalesce example as well

**Comment by matriv** [comment]
Thx @slinkydeveloper! LGTM!
